## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
print(df.describe())


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# Summary of the data frame 
df.info



<bound method DataFrame.info of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  

In [5]:
# Show data frame data types
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [6]:
# Return the number of unique values in each column
df.nunique()

first            360
last             477
job              498
phone            999
email            999
client_id        999
order_id       10000
order_date       150
order_week        23
order_year         2
item_id          500
category           5
subcategory       18
unit_price       490
unit_cost        487
unit_weight      393
qty             2911
line_number       10
dtype: int64

In [7]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [8]:
#Show each category and the number of entries in each
df['category'].value_counts()

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [9]:
#show the number of unique value in each subcategory
df['subcategory'].value_counts()

subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
seating               2664
pens                  2603
logistics             1122
telecom                978
audio visual           732
cabling                655
desks                  510
computers              467
Name: count, dtype: int64

In [10]:
# Filter the DataFrame for rows where the category is 'consumables'
consumables_df = df[df['category'] == 'consumables']

# Count the occurrences of each sub_category within the 'consumables' category
# and display the most common sub_category
most_common_subcategory = consumables_df['subcategory'].value_counts().head(1)

print(most_common_subcategory)


subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [11]:
# Which five clients had the most entries in the data?
top_5_client = df[['client_id']].value_counts().nlargest(5)
print(top_5_client)



client_id
33615        220
66037        211
46820        209
38378        207
24741        207
Name: count, dtype: int64


In [12]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = df['client_id'].value_counts().head(5).index.tolist()
# Sorting a DataFrame by a specific column in descending order
sorted_df = df.sort_values(by='client_id', ascending=False)
print(top_5_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [13]:
# How many total units (the qty column) did the
# client with the most entries order?
sorted_df = df.sort_values(by=['qty'], ascending=[False])
#top_client_id = df['client_id'].value_counts().idxmax()
total_units = df[df['client_id'] == top_client_id]['qty'].sum()
print(f"Total units ordered by the client with the most entries (Client ID: {top_client_id}): {total_units}")
#print(sorted_df.head(5))




NameError: name 'top_client_id' is not defined

In [ ]:
# Store the client ids of those top 5 clients in a list.
top_five_client = df['client_id'].value_counts().nlargest(5)
top_five_client_ids = top_five_client.index.tolist()
print(top_five_client)
#top_five_client_ids  = top_five_client.index.tolist()
#print(top_five_client)

# Filter the DataFrame to include only the data related to the top 5 clients
#top_5_clients_df = df[df['client_id'].isin(top_5_client_ids)]

# Grouping the data by client_id and aggregating the columns
#grouped_df = top_5_clients_df.groupby('client_id').agg({
#      'first': 'first',
#     'last': 'first',
#     'job': 'first',
#     'phone': 'first',
#     'email': 'first',
#     'order_id': 'nunique',
#     'order_date': ['min', 'max'],
#     'qty': 'sum'
# #}).reset_index()

# Renaming the columns for clarity
#grouped_df.columns = ['client_id', 'first', 'last', 'job', 'phone', 'email', 
#                      'total_orders', 'first_order_date', 'last_order_date', 'total_units']

# Sorting the dataframe based on total_units in descending order
#sorted_grouped_df = grouped_df.sort_values(by='total_units', ascending=False)

# Getting the top client's information
#top_client_info = sorted_grouped_df.iloc[0]

# Displaying the sorted grouped DataFrame
#print(sorted_grouped_df)

#print(f"Total units ordered by the client with the most entries (Client ID: {top_client_info['client_id']}): {top_client_info['total_units']}")
#print(top_client_info)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [ ]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
subtotal = df['unit_price'] * df['qty']
print(subtotal)



0        115164.00
1           523.95
2           527.28
3          1056.18
4          3902.00
           ...    
54634      2743.29
54635      9709.73
54636     31188.50
54637       165.76
54638      3103.31
Length: 54639, dtype: float64


In [ ]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def calculate_shipping_price(row):
    shipping_weight = row["unit_weight"] * row["qty"]
    if shipping_weight > 50:
        return shipping_weight * 7
    else:
        return shipping_weight * 10
df["shipping_price"] = df.apply(calculate_shipping_price, axis=1)
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  shipping_price  
0     1096.80     762.71         7.50  105            1         5512.50  
1       24.95      15.09         1.49   21            0          312.90  
2       13.52       7.86         1.68   39            6          458.64  
3       36.42      24.85         1.23   29            3          356.70  
4      195.10     108.17        46.43   20            1         6500.20

In [ ]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['sales_tax_rate'] = 0.0925
df['total_price'] = (df['unit_price'] * df['qty'] + df['shipping_price']) * (1 + df['sales_tax_rate'])
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  shipping_price  sales_tax_rate  \
0     762.71         7.50  105            1         5512.50          0.0925   
1      15.09         1.49   21            0          312.90          0.0925   
2       7.86         1.68   39            6          458.64          0.0925   
3      24.85         1.23   29            3          356.70          0.0925   
4     108.17        46.43   20            1         6500.20          0.0925   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 21 columns]

In [ ]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  shipping_price  sales_tax_rate  \
0         7.50  105            1         5512.50          0.0925   
1         1.49   21            0          312.90          0.0925   
2         1.68   39            6          458.64          0.0925   
3         1.23   29            3          356.70          0.0925   
4        46.43   20            1         6500.20          0.0925   

     total_price      cost  
0  131839.076250  85597.05  
1     914.258625    629.79  
2    1077.117600    765.18  
3    1543.571400   1077.35  
4   11364.403500   8663.60  

[5 rows x 22 columns]

In [ ]:
# Create a column for the profit of each line using line cost and line price
df['unit_profit'] = df['unit_price'] - df['unit_cost']
df['total_profit'] = df['unit_profit'] * df['qty']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  sales_tax_rate    total_price      cost  unit_profit  \
0         5512.50          0.0925  131839.076250  85597.05       334.09   
1          312.90          0.0925     914.258625    629.79         9.86   
2          458.64          0.0925    1077.117600    765.18         5.66   
3          356.70          0.0925    1543.571400   1077.35        11.57   
4         6500.20          0.0925   11364.403500   8663.60        86.93   

   total_profit  
0      35079.45  
1        207.06  
2        220.74  
3        335.53  
4       1738.60  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above
order_ids = [2742071,2173913,6128929]
total_prices = []
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  sales_tax_rate    total_price      cost  unit_profit  \
0         5512.50          0.0925  131839.076250  85597.05       334.09   
1          312.90          0.0925     914.258625    629.79         9.86   
2          458.64          0.0925    1077.117600    765.18         5.66   
3          356.70          0.0925    1543.571400   1077.35        11.57   
4         6500.20          0.0925   11364.403500   8663.60        86.93   

   total_profit  
0      35079.45  
1        207.06  
2        220.74  
3        335.53  
4       1738.60  

[5 rows x 24 columns]

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [ ]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
df['total_price'] = df['unit_price'] * df['qty'] + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  sales_tax_rate  total_price      cost  unit_profit  \
0         5512.50          0.0925    120676.50  85597.05       334.09   
1          312.90          0.0925       836.85    629.79         9.86   
2          458.64          0.0925       985.92    765.18         5.66   
3          356.70          0.0925      1412.88   1077.35        11.57   
4         6500.20          0.0925     10402.20   8663.60        86.93   

   total_profit  
0      35079.45  
1        207.06  
2        220.74  
3        335.53  
4       1738.60  

[5 rows x 24 columns]

In [ ]:
top_5_client_ids = df['client_id'].value_counts().head(5).index.tolist()

# Filter the DataFrame for these top 5 clients, aggregate their data, and sort by total profit
aggregated_data = df[df['client_id'].isin(top_5_client_ids)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'total_profit': 'sum'
}).sort_values(by='total_profit', ascending=False)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  sales_tax_rate  total_price      cost  unit_profit  \
0         5512.50          0.0925    120676.50  85597.05       334.09   
1          312.90          0.0925       836.85    629.79         9.86   
2          458.64          0.0925       985.92    765.18         5.66   
3          356.70          0.0925      1412.88   1077.35        11.57   
4         6500.20          0.0925     10402.20   8663.60        86.93   

   total_profit  
0      35079.45  
1        207.06  
2        220.74  
3        335.53  
4       1738.60  

[5 rows x 24 columns]

In [ ]:
currency_columns = ['total_profit', 'shipping_price', 'total_price']
for column in currency_columns:
    df[column] = df[column] / 1000000  # Convert to millions
    df[column] = df[column].map('${:,.2f}M'.format)  # Format as string with $ sign and M for million
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   shipping_price  sales_tax_rate  total_price      cost unit_profit  \
0          $0.01M          0.0925       $0.12M  85597.05      334.09   
1          $0.00M          0.0925       $0.00M    629.79        9.86   
2          $0.00M          0.0925       $0.00M    765.18        5.66   
3          $0.00M          0.0925       $0.00M   1077.35       11.57   
4          $0.01M          0.0925       $0.01M   8663.60       86.93   

   total_profit  
0        $0.04M  
1        $0.00M  
2        $0.00M  
3        $0.00M  
4        $0.00M  

[5 rows x 24 columns]

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
df = df.sort_values (by = 'total_profit', ascending = False)
df.head()

first     last                                  job  \
15354  Abigail  Johnson                     Financial trader   
15112   Dennis  Hawkins  Social research officer, government   
8761   Randall   Watson                    Ceramics designer   
15413   Lauren    Reese                       Radio producer   
12597   Jordan    Price             Local government officer   

                       phone                                email  client_id  \
15354      799.940.6761x0733                ajohnson5929@king.org      44673   
15112  001-214-888-5838x7410          hawkins.dennis.5195@lee.com      85931   
8761              3478635175  watson.hospital@scott-mcdonald.info      30143   
15413     977.877.5272x11382   late_reese_4081@montoya-chavez.org      29043   
12597           687.750.2141          pricej@donaldson-garcia.com      62227   

       order_id  order_date  order_week  order_year  ... unit_cost  \
15354   6433365  2023-02-15           7        2023  ...    117.06   
15112   7263857  2023-03-05           9        2023  ...    183.84   
8761    3566639  2023-02-21           8        2023  ...     15.09   
15413   5809872  2023-05-20          20        2023  ...    163.92   
12597   3989744  2023-01-23           4        2023  ...    198.38   

      unit_weight     qty  line_number  shipping_price  sales_tax_rate  \
15354        0.12   87378            2          $0.07M          0.0925   
15112        2.21  637318            3          $9.86M          0.0925   
8761         1.49  842581            2          $8.79M          0.0925   
15413        2.05  236016            1          $3.39M          0.0925   
12597       27.51  212123            6         $40.85M          0.0925   

       total_price          cost unit_profit  total_profit  
15354      $19.57M  1.030187e+07      106.05        $9.27M  
15112     $215.98M  1.270239e+08      139.58       $88.96M  
8761       $29.81M  2.150267e+07        9.86        $8.31M  
15413      $48.83M  4.207457e+07       28.61        $6.75M  
12597      $89.58M  8.292949e+07       31.36        $6.65M  

[5 rows x 24 columns]